File: 01-Programming-Languages/Dart/Algorithms/Array/two_sum.dart
/*
Problem: Two Sum
Description: Find two numbers in an array that add up to a target sum.
           Return their indices.

Approach:
- Use a hash map to store complements
- For each element, check if complement (target - element) exists in map

Time Complexity: O(n)
Space Complexity: O(n)

Example:
Input: nums = [2, 7, 11, 15], target = 9
Output: [0, 1]
*/

List<int> twoSum(List<int> nums, int target) {
  Map<int, int> map = {};
  for (int i = 0; i < nums.length; i++) {
    int complement = target - nums[i];
    if (map.containsKey(complement)) {
      return [map[complement]!, i];
    }
    map[nums[i]] = i;
  }
  return [];
}

void main() {
  List<int> result = twoSum([2, 7, 11, 15], 9);
  print('Two Sum indices: $result');
}

File: 01-Programming-Languages/C/Algorithms/Array/two_sum.c
/*
Problem: Two Sum
Given an array of integers and a target, return indices of two numbers that add up to target.

Approach:
- Brute force check all pairs

Time Complexity: O(n^2)
Space Complexity: O(1)

Example:
Input: nums = [2, 7, 11, 15], target = 9
Output: [0, 1]
*/

#include <stdio.h>

void two_sum(int nums[], int n, int target, int *idx1, int *idx2)
{
    for (int i = 0; i < n; i++)
        for (int j = i + 1; j < n; j++)
            if (nums[i] + nums[j] == target)
            {
                *idx1 = i;
                *idx2 = j;
                return;
            }
    *idx1 = -1;
    *idx2 = -1;
}

int main()
{
    int nums[] = {2, 7, 11, 15};
    int n = sizeof(nums) / sizeof(nums[0]);
    int target = 9;
    int i1, i2;
    two_sum(nums, n, target, &i1, &i2);
    printf("Indices: [%d, %d]\n", i1, i2);
    return 0;
}
